### PR TITLE
fix: Fixed an error when the event serializer encountered Exception

### PR DESCRIPTION
**Pull request by liuwwang**
When using langchain, an error was raised and generation could not be recorded

## The following is the duplicate code
```
def test_langfuse():
    from langfuse.callback import CallbackHandler

    langfuse_handler = CallbackHandler(
        public_key="pk-",
        secret_key="sk-",
        host="http://localhost:3000",
        tags=[-2],
        session_id=-2,
        trace_name=-2,
        user_id=-2,
        debug=True
    )

    def chat():
        from langchain_core.output_parsers import StrOutputParser
        from langchain_core.prompts import ChatPromptTemplate
        from langchain_openai import ChatOpenAI

        chat_messages = [
            ("system", "You are a helpful assistant"),
            ("human", "Question: {{question}}")
        ]

        prompt = ChatPromptTemplate.from_messages(
            chat_messages,
            template_format="mustache"
        )

        chain = prompt | ChatOpenAI(
            model_name="qwen-72b",
            openai_api_key="empty",
            openai_api_base="http://localhost:8888/v1",
            max_tokens=100000 # This is intentionally passed here because it will result in errors from the model service
        ) | StrOutputParser()

        content = chain.invoke({"question": "hello"}, config={"callbacks": [langfuse_handler]})
        return content
```
@hassiebp 
I don't know if my modification is in line with your style. If so, please mention it.

### CODE REVIEW

**Comment by liuwwang** [comment]
@hassiebp Thank you for your quick reply, and I have modified it according to the suggested way, and the verification has passed.
Here are the results of the verification on the page:
Snipaste_2024-08-13_17-24-37

**Comment by liuwwang** [comment]
## Here are the problems I encountered before handling the exception

DEBUG:langfuse:consumer is running...
DEBUG:langfuse:Event: on_chain_start, run_id: 5b0d4, parent_run_id: None, inputs: {'question': '你好'}
DEBUG:langfuse:Creating trace id='5b0d4868-ac16-405c-ada2-1b875c72febb' timestamp=datetime.datetime(2024, 8, 13, 9, 26, 49, 983498, tzinfo=datetime.timezone.utc) name='-2' user_id='-2' input={'question': '你好'} output=None session_id='-2' releasNone version=None metadata=None tags=['-2'] public=None
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:Creating span {'id': 'cabbe042-84c7-430e-a9f3-8f716e0b5d48', 'name': 'RunnableSequence', 'start_time': datetime.datetime(2024, 8, 13, 9, 26, 49, 983763, tzinfo=datetime.timezone.utc), 'metadata': None, 'input': {'question': '你好'}'output': None, 'level': None, 'status_message': None, 'version': None, 'end_time': None, 'trace_id': '5b0d4868-ac16-405c-ada2-1b875c72febb'}...
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:Event: on_chain_start, run_id: 0f8c6, parent_run_id: 5b0d4, inputs: {'question': '你好'}
DEBUG:langfuse:item size 318
DEBUG:langfuse:Creating span {'id': 'c35c61d4-20f8-4ef0-b769-f21a7c742d10', 'name': 'ChatPromptTemplate', 'start_time': datetime.datetime(2024, 8, 13, 9, 26, 49, 984756, tzinfo=datetime.timezone.utc), 'metadata': {'tags': ['seq:step:1']}, 'input': {'question': '你好'}, 'output': None, 'level': None, 'status_message': None, 'version': None, 'end_time': None, 'trace_id': '5b0d4868-ac16-405c-ada2-1b875c72febb'}...
DEBUG:langfuse:item size 331
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:Event: on_chain_end, run_id: 0f8c6, parent_run_id: 5b0d4, outputs: messages=[SystemMessage(content='你是个有用的助手'), HumanMessage(content='Question: 你好')]
DEBUG:langfuse:Update span {'id': 'c35c61d4-20f8-4ef0-b769-f21a7c742d10', 'trace_id': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': None, 'start_time': None, 'metadata': None, 'input': None, 'output': ChatPromptValue(messages=[SystemMessage(content='你是个有用的助手'), HumanMessage(content='Question: 你好')]), 'level': None, 'status_message': None, 'version': None, 'end_time': datetime.datetime(2024, 8, 13, 9, 26, 49, 985588, tzinfo=datetime.timezone.utc)}...
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:item size 434
DEBUG:langfuse:Event: on_chat_model_start, run_id: 64e74, parent_run_id: 5b0d4, messages: [[SystemMessage(content='你是个有用的助手'), HumanMessage(content='Question: 你好')]]
DEBUG:langfuse:item size 442
DEBUG:langfuse:Creating generation trace_id='5b0d4868-ac16-405c-ada2-1b875c72febb' name='ChatOpenAI' start_time=datetime.datetime(2024, 8, 13, 9, 26, 49, 986245, tzinfo=datetime.timezone.utc) metadata={'tags': ['seq:step:2']} input=[{'role': 'system', 'content': '你是个有用的助手'}, {'role': 'user', 'content': 'Question: 你好'}] output=None level=None status_message=None parent_observation_id='cabbe042-84c7-430e-a9f3-8f716e0b5d48' version=None id='829ecc9e-7b0e-4f50-b138-d94e3465dd7a' ene completion_start_time=None model='qwen2-72b-instruct-001' model_parameters={'temperature': '0.7', 'max_tokens': 100000} usage=None prompt_name=None prompt_version=None...
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:item size 642
DEBUG:langfuse:Event: on_llm_error, run_id: 64e74, parent_run_id: 5b0d4, error: Error code: 400 - {'error': {'message': "This model's maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", 'type': 'upstream_error', 'param': '400', 'code': 'bad_response_status_code'}}
DEBUG:langfuse:Update generation {'id': '829ecc9e-7b0e-4f50-b138-d94e3465dd7a', 'trace_id': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': None, 'start_time': None, 'metadata': None, 'level': 'ERROR', 'status_message': 'Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}', 'version': None, 'end_time': datetime.datetime(2024, 8, 13, 9, 26, 50, 3498, tzinfo=datetime.timezone.utc), 'completion_start_time': None, 'model': None, 'model_parameters': None, 'input': None, 'output': None, 'usage': None, 'prompt_version': None, 'prompt_name': None}...
trace_id='5b0d4868-ac16-405c-ada2-1b875c72febb' name=None start_time=None metadata=None input=None output=None level=<ObservationLevel.ERROR: 'ERROR'> status_message='Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}' parent_observation_id=None version=None id='829ecc9e-7b0e-4f50-b138-d94e3465dd7a' end_time=datetime.datetime(2024, 8, 13, 9, 26, 50, 3498, tzinfo=datetime.timezone.utc) completion_start_time=None model=None model_parameters=None usage=None prompt_name=None prompt_version=None 99999
DEBUG:langfuse:Update generation {'id': 'a58a5c05-86af-495a-b2cf-7f9e7e08f157', 'type': 'generation-update', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'level': <ObservationLevel.ERROR: 'ERROR'>, 'statusMessage': 'Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}', 'id': '829ecc9e-7b0e-4f50-b138-d94e3465dd7a', 'endTime': datetime.datetime(2024, 8, 13, 9, 26, 50, 3498, tzinfo=datetime.timezone.utc)}}...
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:Event: on_chain_error, run_id: 5b0d4, parent_run_id: None, error: Error code: 400 - {'error': {'message': "This model's maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", 'type': 'upstream_error', 'param': '400', 'code': 'bad_response_status_code'}}
DEBUG:langfuse:Update span {'id': 'cabbe042-84c7-430e-a9f3-8f716e0b5d48', 'trace_id': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': None, 'start_time': None, 'metadata': None, 'input': None, 'output': None, 'level': 'ERROR', 'status_message': 'Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}', 'version': None, 'end_time': datetime.datetime(2024, 8, 13, 9, 26, 50, 3794, tzinfo=datetime.timezone.utc)}...
DEBUG:langfuse:item size 670
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:Update trace {'id': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': None, 'userId': None, 'sessionId': None, 'version': None, 'release': None, 'input': None, 'output': BadRequestError('Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}'), 'metadata': None, 'public': None, 'tags': None}...
DEBUG:langfuse:Applying deterministic sampling to trace_id: 5b0d4868-ac16-405c-ada2-1b875c72febb with rate 1.0
DEBUG:langfuse:item size 664
DEBUG:langfuse:item size 41
Traceback (most recent call last):
  File "/home/liuww/workspace/AI-Platform/tests/langfuse_test.py", line 82, in <module>
    test_langfuse()
  File "/home/liuww/workspace/AI-Platform/tests/langfuse_test.py", line 48, in test_langfuse
    chat()
  File "/home/liuww/workspace/AI-Platform/tests/langfuse_test.py", line 42, in chat
    content = chain.invoke({"question": "你好"}, config={
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_core/runnables/base.py", line 2499, in invoke
    input = step.invoke(
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py", line 154, in invoke
    self.generate_prompt(
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py", line 556, in generate_prompt
    return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py", line 417, in generate
    raise e
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py", line 407, in generate
    self._generate_with_cache(
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py", line 626, in _generate_with_cache
    result = self._generate(
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/langchain_openai/chat_models/base.py", line 438, in _generate
    response = self.client.create(messages=message_dicts, **params)
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/openai/_utils/_utils.py", line 275, in wrapper
    return func(*args, **kwargs)
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/openai/resources/chat/completions.py", line 667, in create
    return self._post(
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/openai/_base_client.py", line 1233, in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/openai/_base_client.py", line 922, in request
    return self._request(
  File "/opt/anaconda3/envs/ai/lib/python3.10/site-packages/openai/_base_client.py", line 1013, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.BadRequestError: Error code: 400 - {'error': {'message': "This model's maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", 'type': 'upstream_error', 'param': '400', 'code': 'bad_response_status_code'}}
DEBUG:langfuse:joining 1 consumer threads
DEBUG:langfuse:~0 items in the Langfuse queue
DEBUG:langfuse:uploading batch of 8 items
DEBUG:langfuse:uploading data: {'batch': [{'id': 'a864804d-38bf-4a93-8c51-46f576fb657d', 'type': 'trace-create', 'body': {'id': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 49, 983498, tzinfo=datetime.timezone.utc), 'name': '-2', 'userId': '-2', 'input': {'question': '你好'}, 'sessionId': '-2', 'tags': ['-2']}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 49, 983719, tzinfo=datetime.timezone.utc)}, {'id': '59261165-ab01-4c38-8f2a-26f6e77e24', 'type': 'span-create', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': 'RunnableSequence', 'startTime': datetime.datetime(2024, 8, 13, 9, 26, 49, 983763, tzinfo=datetime.timezone.utc), 'input': {'question': '你好'}, 'id 'cabbe042-84c7-430e-a9f3-8f716e0b5d48'}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 49, 983929, tzinfo=datetime.timezone.utc)}, {'id': '618d1541-a810-4ceb-bf28-0c38b8702910', 'type': 'span-create', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': 'ChatPromptTemplate', 'startTime': datetime.datetime(2024, 8, 13, 9, 26, 49, 984756, tzinfo=datetime.timezone.utc), 'metadata': {'tags': ['seq:step:1']}, 'input': {'question': '你好'}, 'parentObservationId': 'cabbe0424c7-430e-a9f3-8f716e0b5d48', 'id': 'c35c61d4-20f8-4ef0-b769-f21a7c742d10'}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 49, 985301, tzinfo=datetime.timezone.utc)}, {'id': 'cf46bcb5-ba24-4434-ba0c-16d59c77990e', 'type': 'span-update', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'output': {'messages': [{'content': '你是个有用的助手', 'additional_kwargs': {}}, {'content': 'Question: 你好', 'additional_kwargs': {}}]}, 'id': 'c35c61d4-20f8-4ef0-b769-f21a7c742d10', 'endTtime.datetime(2024, 8, 13, 9, 26, 49, 985588, tzinfo=datetime.timezone.utc)}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 49, 985845, tzinfo=datetime.timezone.utc)}, {'id': '4a9d0126-7c4d-43c8-9286-13710bba2959', 'type': 'generation-create', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'name': 'ChatOpenAI', 'startTime': datetime.datetime(2024, 8, 13, 9, 26, 49, 986245, tzinfo=datetime.timezone.utc), 'metadata': {'tags': ['seq:step:2']}, 'input': [{'role': 'system', 'content': '你是个有用的助手'}, {'role': 'user', 'content': 'Question: 你好'}], 'parentObservationId': 'cabbe042-84c7-430e-a9f3-8f716e0b5d48', 'id': '829ecc9e-7b0e-4f50-b138-d94e3465dd7a', 'model': 'qwen2-72b-instruct-001', 'modelParameters': {'t': '0.7', 'max_tokens': 100000}}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 49, 986520, tzinfo=datetime.timezone.utc)}, {'id': 'a58a5c05-86af-495a-b2cf-7f9e7e08f157', 'type': 'generation-update', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'level': <ObservationLevel.ERROR: 'ERROR'>, 'statusMessage': 'Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}', 'id': '829ecc9e-7b0e-4f50-b138-d94e3465dd7a', 'endTime': datetime.datetime(2024, 8, 13, 9, 26, 50, 3498, tzinfo=datetime.timezone.utc)}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 50, 3736, tzinfo=datetime.timezone.utc)}, {'id': 'd24fe500-8e89-4dae-85ea-0903f8488870', 'type': 'span-update', 'body': {'traceId': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'level': <ObservationLevel.ERROR: 'ERROR'>, 'statusMessage': 'Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}', 'id': 'cabbe042-84c7-430e-a9f3-8f716e0b5d48', 'endTime': datetime.datetime(2024, 8, 13, 9, 26, 50, 3794, tzinfo=datetime.timezone.utc)}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 50, 4150, tzinfo=datetime.timezone.utc)}, {'id': '703e42b3-aefa-46c4-b19e-8e27d98bfd81', 'type': 'trace-create', 'body': {'id': '5b0d4868-ac16-405c-ada2-1b875c72febb', 'output': BadRequestError('Error code: 400 - {\'error\': {\'message\': "This model\'s maximum context length is 32768 tokens. However, you requested 100021 tokens (21 in the messages, 100000 in the completion). Please reduce the length of the messages or completion. (request id: 2024081317264999869545206709718)", \'type\': \'upstream_error\', \'param\': \'400\', \'code\': \'bad_response_status_code\'}}')}, 'timestamp': datetime.datetime(2024, 8, 13, 9, 26, 50, 4665, tzinfo=datetime.timezone.utc)}], 'metadata': {'batch_size': 8, 'sdk_integration': 'langchain', 'sdk_name': 'python', 'sdk_version': '2.43.1', 'public_key': 'pk-lf-c64ed9bd-5a36-4b82-9aa9-2dd853821f58'}}
DEBUG:langfuse:making request: "<not serializable object of type: dict>" to http://localhost:3000/api/public/ingestion
DEBUG:langfuse:received response: {"message":"Invalid request data","errors":["Expected object, received string"]}
ERROR:langfuse:received error response: {'message': 'Invalid request data', 'errors': ['Expected object, received string']}
WARNING:langfuse:Received 400 error by Langfuse server, not retrying: {'message': 'Invalid request data', 'errors': ['Expected object, received string']}
DEBUG:langfuse:successfully uploaded batch of 8 items
DEBUG:langfuse:consumer thread 0 joined

File: langfuse/serializer.py [hunk 1]
```diff
@@ -31,6 +31,9 @@ def default(self, obj: Any):
             # Timezone-awareness check
             return serialize_datetime(obj)
 
+        if isinstance(obj, Exception):
+            return obj.message
```
[the remark this answers is not in the document]
Comment (liuwwang) [reply langfuse/serializer.py]: I modified the code according to your suggestion

**Comment by hassiebp** [comment]
Thanks, did that occur when running the latest version of the Python SDK / from main branch? We have recently joined a fix that should've addressed that issue

**Comment by liuwwang** [comment]
> Thanks, did that occur when running the latest version of the Python SDK / from main branch? We have recently joined a fix that should've addressed that issue

Yes, I have paid attention to the relevant issue and your fix version, and I have also upgraded the SDK and found that it does not solve the problem I am currently experiencing.
That's why I brought up the PR

File: langfuse/serializer.py [hunk 2]
```diff
@@ -31,6 +31,9 @@ def default(self, obj: Any):
             # Timezone-awareness check
             return serialize_datetime(obj)
 
+        if isinstance(obj, Exception):
```
Comment (vegetablest) [inline langfuse/serializer.py:34]: I suggest changing it to this, WDYT?
```
if isinstance(obj, (Exception, KeyboardInterrupt)):
    return f"{type(obj).__name__}: {str(obj)}"
```

**Comment by vegetablest** [comment]
I have the same problem, what's the progress now?

File: langfuse/serializer.py (hunk 2 above)
Comment (liuwwang) [reply langfuse/serializer.py:34]: That makes the code stronger, i will change it.

**Comment by liuwwang** [comment]
> I have the same problem, what's the progress now?

This PR is not merged for the time being, you can wrap the package itself into a package that pip can install to fix this temporarily

**Comment by liuwwang** [comment]
This problem is encountered when the inference engine is vLLM.

**Review by hassiebp — APPROVED** [review]
Thanks for your contribution, @liuwwang!

**Pull request merged**